docs: compile guide (#1575)
AztecProtocol/aztec-packages: docs/docs/dev_docs/contracts/compiling.md

# Compiling contracts

> Please use the [TUTORIAL-TEMPLATE](../../TUTORIAL_TEMPLATE.md) for standalone guides / tutorials.

Reviewer: Nobody likes the template :(

Author: Heh, sorry! I felt it was too extensive for what I was looking for. Maybe having a slightly abridged version?

Reviewer: I was taking inspiration from the uniswap developer guides. They all follow the exact same format, so there's no variance depending on the author of the file. Each page presents info in the same way. But no one's using the template. Maybe devrel can choose to iterate towards a template in future.

> Once you have written your [contracts](../contracts/main.md) in Noir, you will need to compile them into an artifact in order to use them. In this guide we will cover how to do so, both using the CLI and programmatically, as well as generating helper typescript and Noir interfaces for interacting with them.

Reviewer: ```suggestion
Once you have written a [contract](../contracts/main.md) in Noir, you will need to compile it into an [artifact](./abi.md) in order to use it.

In this guide we will cover how to do so, both using the CLI and programmatically.

We'll also cover how to generate a helper [TypeScript interface](#typescript-interface) and a [Noir interface](#noir-interface) for easily interacting with your contract from your typescript app and from other noir contracts, respectively.
```

(I might have got the syntax for internally linking to a markdown header wrong).

## Prerequisites

You will need the Noir build tool `nargo`, which you can install via [`noirup`](https://github.com/noir-lang/noirup). Make sure you install the `aztec` version of nargo:

```bash

> curl -L https://raw.githubusercontent.com/noir-lang/noirup/main/install | bash

Reviewer: Is it best to execute this from the home (`~`) directory?

Author: No need, you should be able to run it from wherever you want

noirup -v aztec
```

:::info
You can re-run `noirup -v aztec` whenever you want to update to the latest version of Noir supported by the Aztec Network.
:::

## Compile using the CLI

> To compile using the Aztec CLI, first install it:

Reviewer: ```suggestion
To compile a contract using the Aztec CLI, first install it:
```

`npm install -g @aztec/cli`

Then run the `compile` command with the path to your Noir contract project:

```

> aztec-cli compile ./src/noir

Reviewer: I just tried this command, myself, and I'm getting `error: unknown command 'compile'`. Is CI not pushing correctly to master? (I had similar problems when trying to follow the quick start guide that you wrote yesterday #1579 )

Author: Indeed, CLI is being released only on tagged releases

> ```

Reviewer: Is this the path the the directory containing `Nargo.toml`, or the path to the `src` directory containing `main.nr`? Perhaps we could make that clearer by using a more descriptive path name?

Reviewer: Oh, and where (which dir) should this commnd be executed from?

Author: Clarified it, and added a reference to the directory structure doc

This will output a JSON artifact for each contract in the project to a `target` folder containing their [ABI](./abi.md), which you can use for deploying or interacting with your contracts.

> ### Typescript

Reviewer: ```suggestion
### Typescript Interfaces
```

You can use the compiler to autogenerate type-safe typescript classes for each of your contracts. These classes define type-safe methods for deploying and interacting with your contract based on their ABI.

To generate them, include a `--typescript` option in the compile command with a path to the target folder for the typescript files:

```

> aztec-cli compile --typescript ./src/ts/contracts ./src/noir

Reviewer: Maybe make the input path name more descriptive (as per another comment).
We might even want to have required named cli parameters for the output-dir and input-dir? (But not for this PR).

Author: > We might even want to have required named cli parameters for the output-dir and input-dir?

Hmm disagree there. If there is a very specific single thing that is the target of a command, I think it makes more sense as a positional argument rather than a named one. In this case, "I want to compile this". As for output dir, I think it's better that it's optional, as long as we're telling the user where things were written to.

Reviewer: It's interesting, because my intuition is that the ordering of the args (as they are ordered currently) is backwards. I would expect "the directory I'm compiling" to be the 1st argument, and "the directory I'm outputting code to" to be the 2nd argument. (Similar to the unix `mv` or `cp` commands).

It was this confusion (relative to my intuition) that led me to propose naming the arguments (to avoid the confusion).

Alternatively, we could copy `solc` and have a positional argument for the source directory, and an `--output` flag for the output directory?

Anyway, this could be a separate issue and PR (if at all). I don't want to hold up this lovely page, given that the current explanation accurately reflects the current code.

Author: > Alternatively, we could copy solc and have a positional argument for the source directory, and an --output flag for the output directory?

That's my proposal here. The source directory is the only thing required, and the main object you are referring to. The output directory is optional, so we can keep it behind an option.

Just to make sure it's clear: in this example `./src/ts/contracts` is **not** a positional argument, but the value of the `typescript` option.

Reviewer: Ohhhh, I misunderstood and assumed --typescript was a flag without any arguments.

```

Example code generated from the [PrivateToken](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/noir-contracts/src/contracts/private_token_contract/src/main.nr) contract:

```ts showLineNumbers
export class PrivateTokenContract extends ContractBase {
  /** Creates a contract instance at the given address. */
  public static async at(address: AztecAddress, wallet: Wallet) { ... }

  /** Creates a tx to deploy a new instance of this contract. */
  public static deploy(rpc: AztecRPC, initial_supply: FieldLike, owner: FieldLike) { ... }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** getBalance(owner: field) */
    getBalance: ((owner: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** mint(amount: field, owner: field) */
    mint: ((amount: FieldLike, owner: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** transfer(amount: field, sender: field, recipient: field) */
    transfer: ((amount: FieldLike, sender: FieldLike, recipient: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
```

> 

Reviewer: Do we want to give an example of how to then make a call to one of these methods? Or maybe link to the section of the docs which explains how to make such calls?

Author: Good point! Added a link to `functions.md#contract-interface`.

Author: Agree. Added a link to dapps/main, which seems the best page we have so far on aztec.js.

To make this easier, the compiler can generate a contract interface struct that exposes a convenience method for each function listed in a given contract ABI. These structs are intended to be used from another contract project that calls into the current one.

To generate them, include a `--interface` option in the compile command with ap ath to the target folder for the generated Noir interface files:

```
aztec-cli compile --interface ../client/src/noir ./src/noir
```

Example code generated from the [PrivateToken](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/noir-contracts/src/contracts/private_token_contract/src/main.nr) contract:

```rust
impl PrivateTokenContractInterface {
  fn at(address: Field) -> Self {
      Self { address }
  }
  
  fn mint(
    self, context: &mut PrivateContext, amount: Field, owner: Field
  ) -> [Field; RETURN_VALUES_LENGTH] {
    let mut serialised_args = [0; 2];
    serialised_args[0] = amount;
    serialised_args[1] = owner;

>     context.call_private_function(self.address, 0x1dc9c3c0, serialised_args)

Reviewer 2: sorry if dumb question, but what are these two `0x...` values?

Reviewer: Hi! :) No, not a dumb question at all! They're 'function selectors'. Much like ethereum, a function can be represented by a 4-byte hex identifier, derived as the hash of the function name and its parameter types. (There's a `compute_function_selector` function somewhere in the codebase (I might have got that function name wrong)).
These hex-values are hard-coded here at the time we auto-generate a contract's interface.
Note: we might change the function selector to be more than 4-bytes in future, to reduce the likelihood of collisions. TBD.

Reviewer: Since the first person to read this doc (Dan) had this question, we should add an explanation (or a hyperlink to an explanation) of what the hex values are.

Reviewer 2: ahh ok great thanks!  got it thank you Michael, good to know we also have function selectors here

Author: Added a comment with the explanation in the code.

  }
  

  fn transfer(
    self, context: &mut PrivateContext, amount: Field, sender: Field, recipient: Field
  ) -> [Field; RETURN_VALUES_LENGTH] {
    let mut serialised_args = [0; 3];
    serialised_args[0] = amount;
    serialised_args[1] = sender;
    serialised_args[2] = recipient;

    context.call_private_function(self.address, 0xdcd4c318, serialised_args)
  }
}

```

> 

Reviewer: Do we want to give an example of how one would make a call to one of these interface methods, from some other Noir contract?

Author: Added a link to functions.md#contract-interface

:::info
At the moment, the compiler generates these interfaces from already compiled ABIs, and not from source code. This means that you should not import a generated interface from within the same project as its source contract, or you risk circular references.
:::

> ## Access the compiler using nodejs

Reviewer: ```suggestion
## Compile using nodejs
```
(Consistency of naming vs the `Compile using the CLI` subheading)
